### ISSUE TITLE
Playing track skips when new track loaded.

### ISSUE THREAD

**Issue by 3xrp1** [pideck/pideck#107]
Hi, when I load a new track the pitch and speed on the playing track spikes for <1s. Any ideas on what could be causing this?

uname -a
pideck 4.14.39-v7+ #1112 SMP Sat May 5 12:01:33 BST 2018 armv7l GNU/Linux

cat /proc/device-tree/model
Raspberry Pi 3 Model B Rev 1.2

I'm using an M-Audio connective which was woking fine with the xwax installed on another computer.

I'm running xwax with 2 decks, my startup cmd is 

sudo xwax -t serato_2a -l /media/usb0 -a hw:1,0 -a hw:1,1

Flash drive is
sandisk cruzer glide sdcz20-061

blkid (run on another computer since lsusb and blkid don't work in pideck)

/dev/sda1: Label="UBUNTU-BUDG" UUID="880A-AA3D" Type="vfat" Partuuid="05a4789b01"

appears as /media/usb0 on pideck

Power supply
Input:100-240V~
50/60 Hz 0.35A
Output: 5.25V[line with three dots underneath]2400mA

I'm using a wireless keyboard and mouse.

I remember reading a thread somewhere (might have been on xwax forum) where there was an issue with skipping I think to do with sample rate and someone ended up using this piece of software that was sort of like an oscilliscope to identify what was happening? The issue was solved it was some sort of intererence I think. Anyways I can't find it and want to reread so if anyone also read it and remembers it more clearly please link! tysm

**Comment by danielhjames** [comment]
Hi STemR,If you are writing your own start-up command for the Conectiv rather than using the PiDeck scripts, please try adding:-m 5 -r 44100before the first -a entry so that it reads:xwax -l /media/usb0 -m 5 -r 44100 -a hw:1,0 -a hw:1,1This is because the Conectiv we tested doesn't work natively at xwax's default 48000 Hz sample rate, and a buffer of 4ms is OK for some USB devices but the Conectiv needs 5ms to work without under-runs in our tests. You shouldn't need to use sudo or to specify the default timecode.Alternatively, please see the recent commits in https://github.com/pideck/pideck/commits/master in which we have added auto detection for the Conectiv.Please let us know how you get on!Daniel
null

**Comment by 3xrp1** [comment]
Hm, I tried this with -m up to 15ms and I'm still getting the same little distorted skips when I load a track. I'll check out the autodetection. I'm using my own startup command because xwax isn't starting on boot so I'll try and fix that and see if it helps.

**Comment by stjohn909** [comment]
Confirming this.  Fresh 0.2 image, same Pi hardware as OP, but Rane SL-1 soundcard instead.  Brief skip and warble when loading a new track into either deck.    In my case, I added all the latest pideck scripts to /usr/share/pideck from this repo, but still getting the skips.  it's not as bad if you do it on the beat, but it's there.  Change of buffer size has no effect.

USB drive is a SanDisk Extreme USB3.0 64GB stick.

**Comment by GuusR** [comment]
Same problem here with two Behringer UFO 202's. I've tried lowering to 44100 but that made everything sound a bit 'glitchy' and the stutter problem when loading a new track is still present.
Anyone got an idea how this can be solved?

**Comment by GuusR** [comment]
I also noticed that the skipping only happens after the first two loads. 
I.e. after startup I load to the left deck and start playing. Then when I load a song to the (still empty) right deck there is no stutter/skip. After this any subsequent track loading will make the playing deck skip.
So it seems the pi has a hard time loading a track when both decks are loaded but seems to be okay if only one deck has been loaded.

Is it possible to add a "clear deck" command right before the "load deck" command? I couldn't find a keyboard command for this, but there must be a way to do this right?

**Comment by danielhjames** [comment]
Hi @GuusR that is useful feedback, thanks! I don't know of a command to clear a deck, but at best that would be a workaround and might not solve the issue. I will ask on the xwax-devel mailing list for any clues.

**Comment by GuusR** [comment]
Thanks, I really hope this can be solved!

**Comment by Supercafone** [comment]
For what it's worth, I have the same issue on my pi 3b+ using a mixvibes u-mix44. Have not gotten around to looking into it so no additional information to add so far.

**Comment by kretek77** [comment]
Same here with a Pi 3B+ and a Rane SL-1. Has anyone fixed it yet? Perhaps by overclocking?

**Comment by johansmolinski** [comment]
I have implemented a fix for this problem. Refer to issue #89 for details.

**Comment by GuusR** [comment]
@johansmolinski Any updates on the fix? I can't wait to try it on my device! :)

**Comment by johansmolinski** [comment]
No. I have asked for feedback with no response, so my proposed fix is still in my branch for anyone to try. :)

**Comment by kretek77** [comment]
Sorry for that. I somehow succeeded in building your xwax code on the 0.2 pideck image and it also did work. The GUI was not that Pi Touchscreen friendly, but I guess one could change that if one would know how to.

Recently, I tried to install your fix again, but failed, because I forgot how to install the required dependencies. As you might suspect I am by no means an expert in linux, so would it be possible for someone to
a) either describe the specific building procedure on the pideck image or
b) upload a .deb or
c) perhaps offer a pre-built pideck image with your fix?

**Comment by danielhjames** [comment]
Hi @kretek77 the GUI from Johan's branch would not have the small screen tweaks we made in https://github.com/xwax/xwax/commit/ca456e6eb1c176bd9e64682d75d71b7c1f6130fa
The build instructions are at https://github.com/pideck/xwax/blob/pideck/README.md

**Comment by danielhjames** [comment]
Hi @johansmolinski I recall that you mentioned xwax does not free up memory with your patch. For PiDeck purposes, as we don't intend to restart xwax during a DJ set (i.e. between reboots), and there are no other desktop programs running, do you think it would be OK just to let the operating system reclaim the memory in the event of xwax quitting unexpectedly?

There is a discussion on this topic here: https://stackoverflow.com/questions/2213627/when-you-exit-a-c-application-is-the-malloc-ed-memory-automatically-freed

**Comment by Supercafone** [comment]
I've been trying build the patched version, but I'm stuck at make giving the following output:
`./mkversion -r
./mkversion: Version number is not known
cc -O3 -Wall `sdl-config --cflags` -MMD  -DWITH_ALSA -DWITH_JACK -DEXECDIR=\"/usr/libexec\" -DVERSION=\"\"  -c -o xwax.o xwax.c
cc -O3 -Wall `sdl-config --cflags` -MMD  -c -o interface.o interface.c
interface.c: In function 'draw_clock':
interface.c:213:29: warning: 'sprintf' may write a terminating nul past the end of the destination [-Wformat-overflow=]
     sprintf(buf, "%02d:%02d.", minutes, seconds);
                             ^
interface.c:213:5: note: 'sprintf' output between 7 and 9 bytes into a destination of size 8
     sprintf(buf, "%02d:%02d.", minutes, seconds);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc -O3 -pthread  xwax.o controller.o cues.o deck.o device.o dummy.o excrate.o external.o index.o interface.o library.o listbox.o lut.o player.o realtime.o rig.o selector.o status.o thread.o timecoder.o track.o alsa.o dicer.o midi.o jack.o  `sdl-config --libs` -lSDL_ttf  -lasound -ljack -lm -o xwax
/usr/bin/ld: cannot find -lSDL_ttf
collect2: error: ld returned 1 exit status
make: *** [<builtin>: xwax] Error 1`

I've tried googling for a solution, but it appears to be a bit over my head as this is my first time building something from source. To be sure some extra information on how I got here:

It seemed some of the dependencies did not install correctly by copying the commands from the instructions, possibly because the partition was full, which I solved by increasing the partition size on the sd. But after some repetitions of installing the missing ones that seemed to be sorted out.  

I might have made some obvious mistake along the way. Is it correct to copy the files to some folder on the pideck and then execute the instructions in the read-me? I'm also not completely sure what the result is supposed to be, is this overwriting the pideck-version of xwax, or installing the patched version next to it?

Thanks in advance!

**Comment by danielhjames** [comment]
Hi @Supercafone the probable issue is that you have font libraries missing. The clue is in this part:

`cannot find -lSDL_ttf`

This suggests the package `libsdl-ttf2.0-dev` is missing. Once build has completed you can run this new copy from the local directory using a command starting with `./xwax` rather than `xwax` which would run the previously installed binary `/usr/bin/xwax` from the regular package.

**Comment by Supercafone** [comment]
Thanks again @danielhjames! I figured as much, but the package was installed and I wasn't able to fix it, possibly I broke something before in my attempts to get the GPIO running. I started with a clean image and got it built, it seems to work: no more glitches, so thanks a bunch @johansmolinski, I will play around with it some more tomorrow and report any issues I might encounter! My pi is a 3b+, just in case that is useful information.

For the other noobs wanting to try, some things I found out along the way:
- Building in the root folder will make it impossible to start it from the pideck script by changing 'xwax' to the location of the new file in the start up command. This works fine when it is on the second, bigger partition.
- Installing of libjack-dev will uninstall pideck, holding it with apt didn't work, so I skipped that one and built without --enable-jack, as I'm not using jack.
- Unrelated, but possibly useful: changing buffer size to 64 and period size to 4 in the asound file works for me at a 3 ms buffer on my mixvides umix-44 (again unrelated, but can I submit my cards_autodetect addition and asound.conf for this card for a future version somewhere, without messing anything else up?)
(Edited values, it was a bit late yesterday)

**Comment by danielhjames** [comment]
Hi @Supercafone thanks for the tips! You can make a pull request here on GitHub for your improvements, or if you're not familiar with how to do that, please attach your files to this ticket and we will merge them.
